File: frontend/src/simulation/HolographicMemory.js
/**
 * PROTEUS Holographic Memory System
 * Each point contains information about the whole experience
 * Robust to damage, allows superposition of memories
 */

export class HolographicMemory {
  constructor(size = 1000, parent1 = null, parent2 = null) {
    this.size = size;
    
    if (parent1 && parent2) {
      // Inherit through interference pattern
      this.interferencePattern = this.createInterference(parent1, parent2);
      this.phaseRelations = this.blendPhases(parent1.phaseRelations, parent2.phaseRelations);
    } else if (parent1) {
      // Single parent - mutate slightly
      this.interferencePattern = this.mutatePattern(parent1.interferencePattern);
      this.phaseRelations = this.mutatePattern(parent1.phaseRelations);
    } else {
      // Genesis - random initialization
      this.interferencePattern = new Float32Array(size);
      this.phaseRelations = new Float32Array(size);
      this.initializeRandom();
    }
    
    // Compressed history of critical moments
    this.criticalMoments = [];
    this.memoryStrength = 1.0;
  }
  
  initializeRandom() {
    for (let i = 0; i < this.size; i++) {
      this.interferencePattern[i] = Math.random() * 2 - 1;
      this.phaseRelations[i] = Math.random() * Math.PI * 2;
    }
  }
  
  // Create interference pattern from two parents
  createInterference(parent1, parent2) {
    const pattern = new Float32Array(this.size);
    const p1 = parent1.interferencePattern;
    const p2 = parent2.interferencePattern;
    
    for (let i = 0; i < this.size; i++) {
      // Not simple average - true wave interference
      const phase1 = parent1.phaseRelations[i];
      const phase2 = parent2.phaseRelations[i];
      
      // Constructive and destructive interference
      pattern[i] = p1[i] * Math.cos(phase1) + 
                   p2[i] * Math.cos(phase2) + 
                   2 * Math.sqrt(Math.abs(p1[i] * p2[i])) * Math.cos(phase1 - phase2);
      
      // Normalize
      pattern[i] = Math.tanh(pattern[i] * 0.5);
    }
    
    return pattern;
  }
  
  blendPhases(phases1, phases2) {
    const blended = new Float32Array(this.size);
    for (let i = 0; i < this.size; i++) {
      // Circular mean for phases
      const x = Math.cos(phases1[i]) + Math.cos(phases2[i]);
      const y = Math.sin(phases1[i]) + Math.sin(phases2[i]);
      blended[i] = Math.atan2(y, x);
    }
    return blended;
  }
  
  mutatePattern(pattern, mutationRate = 0.1) {
    const mutated = new Float32Array(pattern);
    for (let i = 0; i < pattern.length; i++) {
      if (Math.random() < mutationRate) {
        mutated[i] += (Math.random() - 0.5) * 0.2;
        mutated[i] = Math.tanh(mutated[i]);
      }
    }
    return mutated;
  }
  
  // Encode an experience into the hologram
  encode(experience) {
    // Extract frequency components from experience
    const { trajectory, importance, emotionalValue, survivalValue } = experience;
    
    // Simple frequency extraction (in real implementation, use FFT)
    const frequencies = this.extractFrequencies(trajectory);
    
    // Update interference pattern
    const learningRate = 0.1 * importance;
    for (let i = 0; i < this.size; i++) {
      // Each point gets contribution from ALL frequencies
      let contribution = 0;
      for (let f = 0; f < frequencies.length; f++) {
        contribution += frequencies[f] * Math.cos(f * i * 0.01 + this.phaseRelations[i]);
      }
      
      // Update with momentum
      this.interferencePattern[i] = 
        (1 - learningRate) * this.interferencePattern[i] + 
        learningRate * Math.tanh(contribution);
    }
    
    // Store critical moments
    if (importance > 0.7) {
      this.criticalMoments.push({
        timestamp: Date.now(),
        topology: this.extractTopology(experience),
        impact: importance,
        type: experience.type
      });
      
      // Keep only most recent critical moments
      if (this.criticalMoments.length > 20) {
        this.criticalMoments.shift();
      }
    }
  }
  
  extractFrequencies(trajectory) {
    // Simplified frequency extraction
    const frequencies = new Float32Array(10);
    
    if (!trajectory || trajectory.length === 0) {
      return frequencies;
    }
    
    // Extract basic frequency components
    for (let f = 0; f < 10; f++) {
      let sum = 0;
      for (let i = 0; i < trajectory.length; i++) {
        const point = trajectory[i];
        sum += Math.sqrt(point.x * point.x + point.y * point.y) * 
               Math.cos(2 * Math.PI * f * i / trajectory.length);
      }
      frequencies[f] = sum / trajectory.length;
    }
    
    return frequencies;
  }
  
  extractTopology(experience) {
    const { trajectory, predatorDistance, nutrientDirection } = experience;
    
    // Extract topological invariants
    return {
      curvature: this.calculateCurvature(trajectory),
      windingNumber: this.calculateWindingNumber(trajectory),
      persistence: this.calculatePersistence(trajectory),
      danger: predatorDistance ? 1 / predatorDistance : 0,
      attraction: nutrientDirection ? Math.atan2(nutrientDirection.y, nutrientDirection.x) : 0
    };
  }
  
  calculateCurvature(trajectory) {
    if (!trajectory || trajectory.length < 3) return 0;
    
    // Average curvature along path
    let totalCurvature = 0;
    for (let i = 1; i < trajectory.length - 1; i++) {
      const p1 = trajectory[i - 1];
      const p2 = trajectory[i];
      const p3 = trajectory[i + 1];
      
      const v1 = { x: p2.x - p1.x, y: p2.y - p1.y };
      const v2 = { x: p3.x - p2.x, y: p3.y - p2.y };
      
      const cross = v1.x * v2.y - v1.y * v2.x;
      const dot = v1.x * v2.x + v1.y * v2.y;
      
      totalCurvature += Math.atan2(cross, dot);
    }
    
    return totalCurvature / (trajectory.length - 2);
  }
  
  calculateWindingNumber(trajectory) {
    if (!trajectory || trajectory.length < 2) return 0;
    
    let totalAngle = 0;
    for (let i = 0; i < trajectory.length - 1; i++) {
      const p1 = trajectory[i];
      const p2 = trajectory[i + 1];
      
      const angle1 = Math.atan2(p1.y, p1.x);
      const angle2 = Math.atan2(p2.y, p2.x);
      
      let dAngle = angle2 - angle1;
      while (dAngle > Math.PI) dAngle -= 2 * Math.PI;
      while (dAngle < -Math.PI) dAngle += 2 * Math.PI;
      
      totalAngle += dAngle;
    }
    
    return totalAngle / (2 * Math.PI);
  }
  
  calculatePersistence(trajectory) {
    // Simplified persistence - how long the organism maintained a direction
    if (!trajectory || trajectory.length < 2) return 0;
    
    let persistence = 0;
    let currentDirection = null;
    
    for (let i = 1; i < trajectory.length; i++) {
      const dx = trajectory[i].x - trajectory[i - 1].x;
      const dy = trajectory[i].y - trajectory[i - 1].y;
      const direction = Math.atan2(dy, dx);
      
      if (currentDirection !== null) {
        const similarity = Math.cos(direction - currentDirection);
        persistence += similarity > 0.8 ? 1 : 0;
      }
      currentDirection = direction;
    }
    
    return persistence / trajectory.length;
  }
  
  // Recall behavior from holographic memory
  recall(context = {}) {
    const { danger, hunger, energy } = context;
    
    // Use interference pattern to generate behavior
    const behavior = {
      explorationTendency: 0,
      cautionLevel: 0,
      forageIntensity: 0,
      socialAttraction: 0
    };
    
    // Different regions of the hologram encode different behaviors
    const regions = {
      exploration: { start: 0, end: 250 },
      caution: { start: 250, end: 500 },
      foraging: { start: 500, end: 750 },
      social: { start: 750, end: 1000 }
    };
    
    // Extract behavioral tendencies from interference pattern
    for (const [behaviorType, region] of Object.entries(regions)) {
      let sum = 0;
      for (let i = region.start; i < region.end; i++) {
        sum += this.interferencePattern[i] * Math.cos(this.phaseRelations[i]);
      }
      behavior[behaviorType + 'Tendency'] = Math.tanh(sum / (region.end - region.start));
    }
    
    // Modulate by current context
    if (danger > 0.5) {
      behavior.cautionLevel *= 2;
      behavior.explorationTendency *= 0.5;
    }
    
    if (hunger > 0.7) {
      behavior.forageIntensity *= 2;
    }
    
    return behavior;
  }
  
  // Calculate similarity between two memories
  similarity(other) {
    let correlation = 0;
    for (let i = 0; i < this.size; i++) {
      correlation += this.interferencePattern[i] * other.interferencePattern[i];
    }
    return correlation / this.size;
  }
  
  // Recover from partial damage
  recover(damageRatio = 0.5) {
    // Even with 50% damage, holographic memory can recover
    const damaged = [...this.interferencePattern];
    const damagedIndices = new Set();
    
    // Simulate damage
    for (let i = 0; i < this.size * damageRatio; i++) {
      const idx = Math.floor(Math.random() * this.size);
      damaged[idx] = 0;
      damagedIndices.add(idx);
    }
    
    // Holographic recovery using undamaged parts
    for (const idx of damagedIndices) {
      let reconstructed = 0;
      let count = 0;
      
      // Each undamaged point contributes to reconstruction
      for (let i = 0; i < this.size; i++) {
        if (!damagedIndices.has(i)) {
          const phase = this.phaseRelations[i] - this.phaseRelations[idx];
          reconstructed += damaged[i] * Math.cos(phase);
          count++;
        }
      }
      
      if (count > 0) {
        this.interferencePattern[idx] = reconstructed / count;
      }
    }
  }
  
  compress() {
    // Return only the most significant frequencies
    const compressed = {
      dominantFrequencies: [],
      criticalMoments: this.criticalMoments.slice(-5),
      signature: new Float32Array(20)
    };
    
    // Extract dominant frequencies
    for (let f = 0; f < 20; f++) {
      let power = 0;
      for (let i = 0; i < this.size; i++) {
        power += this.interferencePattern[i] * Math.cos(f * i * 0.01);
      }
      compressed.signature[f] = power / this.size;
    }
    
    return compressed;
  }
}

File: backend/proteus/genetics/holographic_memory.py
"""
Holographic Memory System for PROTEUS
Each point contains information about the whole experience
"""

import numpy as np
from numba import jit, prange
from typing import Dict, List, Tuple, Optional
import json

class HolographicMemory:
    """
    Holographic memory that stores experiences as interference patterns
    Robust to damage and allows experience inheritance
    """
    
    def __init__(self, size: int = 1000, parent1=None, parent2=None):
        self.size = size
        
        if parent1 and parent2:
            # Inherit through interference
            self.interference_pattern = self._create_interference(parent1, parent2)
            self.phase_relations = self._blend_phases(
                parent1.phase_relations, 
                parent2.phase_relations
            )
        elif parent1:
            # Single parent - mutate
            self.interference_pattern = self._mutate_pattern(parent1.interference_pattern)
            self.phase_relations = self._mutate_pattern(parent1.phase_relations)
        else:
            # Genesis
            self.interference_pattern = np.random.randn(size).astype(np.float32)
            self.phase_relations = np.random.random(size).astype(np.float32) * 2 * np.pi
            
        # Compressed history
        self.critical_moments = []
        self.memory_strength = 1.0
        
    @staticmethod
    @jit(nopython=True)
    def _create_interference_jit(p1, p2, phase1, phase2):
        """JIT-compiled interference creation"""
        size = len(p1)
        pattern = np.zeros(size, dtype=np.float32)
        
        for i in range(size):
            # Wave interference
            pattern[i] = (p1[i] * np.cos(phase1[i]) + 
                         p2[i] * np.cos(phase2[i]) + 
                         2 * np.sqrt(np.abs(p1[i] * p2[i])) * 
                         np.cos(phase1[i] - phase2[i]))
            # Normalize
            pattern[i] = np.tanh(pattern[i] * 0.5)
            
        return pattern
        
    def _create_interference(self, parent1, parent2):
        """Create interference pattern from parents"""
        return self._create_interference_jit(
            parent1.interference_pattern,
            parent2.interference_pattern,
            parent1.phase_relations,
            parent2.phase_relations
        )
        
    @staticmethod
    @jit(nopython=True)
    def _blend_phases_jit(phases1, phases2):
        """JIT-compiled phase blending"""
        size = len(phases1)
        blended = np.zeros(size, dtype=np.float32)
        
        for i in range(size):
            x = np.cos(phases1[i]) + np.cos(phases2[i])
            y = np.sin(phases1[i]) + np.sin(phases2[i])
            blended[i] = np.arctan2(y, x)
            
        return blended
        
    def _blend_phases(self, phases1, phases2):
        """Blend phase relations from parents"""
        return self._blend_phases_jit(phases1, phases2)
        
    @staticmethod
    @jit(nopython=True)
    def _mutate_pattern_jit(pattern, rate):
        """JIT-compiled pattern mutation"""
        mutated = pattern.copy()
        size = len(pattern)
        
        for i in range(size):
            if np.random.random() < rate:
                mutated[i] += (np.random.random() - 0.5) * 0.2
                mutated[i] = np.tanh(mutated[i])
                
        return mutated
        
    def _mutate_pattern(self, pattern, rate=0.1):
        """Mutate a pattern"""
        return self._mutate_pattern_jit(pattern, rate)
        
    def encode(self, experience: Dict):
        """Encode experience into holographic memory"""
        # Extract features
        importance = experience.get('importance', 0.5)
        trajectory = experience.get('trajectory', [])
        
        # Extract frequencies (simplified)
        frequencies = self._extract_frequencies(trajectory)
        
        # Update interference pattern
        self._update_pattern(frequencies, importance)
        
        # Store critical moments
        if importance > 0.7:
            self.critical_moments.append({
                'timestamp': experience.get('timestamp', 0),
                'topology': self._extract_topology(experience),
                'impact': importance,
                'type': experience.get('type', 'unknown')
            })
            
            # Keep only recent
            if len(self.critical_moments) > 20:
                self.critical_moments.pop(0)
                
    @staticmethod
    @jit(nopython=True)
    def _extract_frequencies_jit(trajectory_x, trajectory_y):
        """Extract frequency components from trajectory"""
        if len(trajectory_x) == 0:
            return np.zeros(10, dtype=np.float32)
            
        frequencies = np.zeros(10, dtype=np.float32)
        n = len(trajectory_x)
        
        for f in range(10):
            sum_val = 0.0
            for i in range(n):
                magnitude = np.sqrt(trajectory_x[i]**2 + trajectory_y[i]**2)
                sum_val += magnitude * np.cos(2 * np.pi * f * i / n)
            frequencies[f] = sum_val / n
            
        return frequencies
        
    def _extract_frequencies(self, trajectory):
        """Extract frequencies from trajectory"""
        if not trajectory:
            return np.zeros(10, dtype=np.float32)
            
        # Convert to numpy arrays
        x = np.array([p['x'] for p in trajectory], dtype=np.float32)
        y = np.array([p['y'] for p in trajectory], dtype=np.float32)
        
        return self._extract_frequencies_jit(x, y)
        
    @jit(nopython=True)
    def _update_pattern_jit(pattern, phases, frequencies, learning_rate):
        """JIT-compiled pattern update"""
        size = len(pattern)
        
        for i in range(size):
            contribution = 0.0
            for f in range(len(frequencies)):
                contribution += frequencies[f] * np.cos(f * i * 0.01 + phases[i])
                
            pattern[i] = ((1 - learning_rate) * pattern[i] + 
                         learning_rate * np.tanh(contribution))
                         
    def _update_pattern(self, frequencies, importance):
        """Update interference pattern with new experience"""
        learning_rate = 0.1 * importance
        self._update_pattern_jit(
            self.interference_pattern,
            self.phase_relations,
            frequencies,
            learning_rate
        )
        
    def _extract_topology(self, experience):
        """Extract topological invariants"""
        trajectory = experience.get('trajectory', [])
        
        return {
            'curvature': self._calculate_curvature(trajectory),
            'winding': self._calculate_winding(trajectory),
            'persistence': len(trajectory),
            'danger': 1.0 / experience.get('predator_distance', 1000),
            'attraction': experience.get('nutrient_direction', 0)
        }
        
    def _calculate_curvature(self, trajectory):
        """Calculate average path curvature"""
        if len(trajectory) < 3:
            return 0.0
            
        total_curvature = 0.0
        for i in range(1, len(trajectory) - 1):
            p1 = trajectory[i-1]
            p2 = trajectory[i]
            p3 = trajectory[i+1]
            
            v1 = (p2['x'] - p1['x'], p2['y'] - p1['y'])
            v2 = (p3['x'] - p2['x'], p3['y'] - p2['y'])
            
            cross = v1[0] * v2[1] - v1[1] * v2[0]
            dot = v1[0] * v2[0] + v1[1] * v2[1]
            
            total_curvature += np.arctan2(cross, dot)
            
        return total_curvature / (len(trajectory) - 2)
        
    def _calculate_winding(self, trajectory):
        """Calculate winding number"""
        if len(trajectory) < 2:
            return 0.0
            
        total_angle = 0.0
        for i in range(len(trajectory) - 1):
            p1 = trajectory[i]
            p2 = trajectory[i + 1]
            
            angle1 = np.arctan2(p1['y'], p1['x'])
            angle2 = np.arctan2(p2['y'], p2['x'])
            
            d_angle = angle2 - angle1
            while d_angle > np.pi:
                d_angle -= 2 * np.pi
            while d_angle < -np.pi:
                d_angle += 2 * np.pi
                
            total_angle += d_angle
            
        return total_angle / (2 * np.pi)
        
    def recall(self, context: Dict) -> Dict:
        """Recall behavior from memory given context"""
        danger = context.get('danger', 0)
        hunger = context.get('hunger', 0)
        energy = context.get('energy', 1)
        
        # Extract behavioral tendencies from pattern
        behavior = self._extract_behavior()
        
        # Modulate by context
        if danger > 0.5:
            behavior['caution_level'] *= 2
            behavior['exploration_tendency'] *= 0.5
            
        if hunger > 0.7:
            behavior['forage_intensity'] *= 2
            
        return behavior
        
    @staticmethod
    @jit(nopython=True)
    def _extract_behavior_jit(pattern):
        """Extract behaviors from interference pattern"""
        # Different regions encode different behaviors
        exploration = np.tanh(np.mean(pattern[0:250]))
        caution = np.tanh(np.mean(pattern[250:500]))
        foraging = np.tanh(np.mean(pattern[500:750]))
        social = np.tanh(np.mean(pattern[750:1000]))
        
        return exploration, caution, foraging, social
        
    def _extract_behavior(self):
        """Extract behavioral tendencies"""
        exp, caut, forage, social = self._extract_behavior_jit(
            self.interference_pattern
        )
        
        return {
            'exploration_tendency': exp,
            'caution_level': caut,
            'forage_intensity': forage,
            'social_attraction': social
        }
        
    def similarity(self, other) -> float:
        """Calculate similarity with another memory"""
        return np.corrcoef(
            self.interference_pattern, 
            other.interference_pattern
        )[0, 1]
        
    def compress(self) -> Dict:
        """Compress memory for storage"""
        # Extract dominant frequencies
        fft = np.fft.fft(self.interference_pattern)
        power = np.abs(fft[:20])
        
        return {
            'dominant_frequencies': power.tolist(),
            'critical_moments': self.critical_moments[-5:],
            'signature': power.tolist()
        }
        
    def to_json(self) -> str:
        """Serialize to JSON"""
        return json.dumps({
            'size': self.size,
            'pattern': self.interference_pattern.tolist()[:100],  # Sample
            'phases': self.phase_relations.tolist()[:100],
            'strength': self.memory_strength,
            'critical_moments': self.critical_moments
        })
        
    @classmethod
    def from_json(cls, data: str):
        """Deserialize from JSON"""
        d = json.loads(data)
        memory = cls(size=d['size'])
        # Restore what we can
        memory.memory_strength = d['strength']
        memory.critical_moments = d['critical_moments']
        return memory